Sources/SwiftBasicFormat/BasicFormat.swift

```diff
@@ -143,35 +143,50 @@ open class BasicFormat: SyntaxRewriter {
     return node.requiresIndent
   }
 
-  /// Whether a leading newline on `token` should be added.
-  open func requiresLeadingNewline(_ token: TokenSyntax) -> Bool {
-    // We don't want to add newlines inside string interpolation
-    if isInsideStringInterpolation(token) {
+  open func requiresNewline(between first: TokenSyntax?, and second: TokenSyntax?) -> Bool {
+    // We don't want to add newlines inside string interpolation.
+    // When first or second `TokenSyntax` is a multiline quote we want special handling
+    // event it's inside a string interpolation, because it still requires newline
+    // after open quote and before close quote.
+    if let first,
+      first.tokenKind != .multilineStringQuote,
+      second?.tokenKind != .multilineStringQuote,
+      isInsideStringInterpolation(first)
+    {
       return false
-    }
-
-    if token.requiresLeadingNewline {
-      return true
-    }
-
-    var ancestor: Syntax = Syntax(token)
-    while let parent = ancestor.parent {
-      ancestor = parent
-      if ancestor.position != token.position || ancestor.firstToken(viewMode: viewMode) != token {
-        break
-      }
-      if let ancestorsParent = ancestor.parent, childrenSeparatedByNewline(ancestorsParent) {
+    } else if let second {
+      if second.requiresLeadingNewline {
         return true
       }
-      switch ancestor.keyPathInParent {
-      case \IfConfigClauseSyntax.elements:
-        return true
-      default:
-        break
+
+      var ancestor: Syntax = Syntax(second)
+      while let parent = ancestor.parent {
+        ancestor = parent
+        if ancestor.position != second.position || ancestor.firstToken(viewMode: viewMode) != second {
+          break
+        }
+        if let ancestorsParent = ancestor.parent, childrenSeparatedByNewline(ancestorsParent) {
+          return true
+        }
+        switch ancestor.keyPathInParent {
+        case \IfConfigClauseSyntax.elements:
+          return true
+        default:
+          break
+        }
       }
     }
 
-    return false
+    switch (first?.tokenKind, second?.tokenKind) {
+    case (.multilineStringQuote, .backslash),
+      (.multilineStringQuote, .multilineStringQuote),
+      (.multilineStringQuote, .stringSegment),
+      (.stringSegment, .multilineStringQuote),
+      (.rightParen, .multilineStringQuote):
```

Comment:
Just to follow the paradigm of `tokenSeparatedByNewline`, I think it would be good to add comment to at least part of them (it took me a bit of thinking to see why `multilineStringQuote` and `backslash` need to be separated by a newline, only to realize that it’s because of a string interpolation segment starting a multi-line string literal

```suggestion
    case (.multilineStringQuote, .backslash), // string interpolation segment inside a multi-line string literal
      (.multilineStringQuote, .multilineStringQuote), // empty multi-line string literal
      (.multilineStringQuote, .stringSegment), // segment starting a multi-line string literal
      (.stringSegment, .multilineStringQuote), // ending a multi-line string literal on a new line
      (.rightParen, .multilineStringQuote): // ending a multi-line string literal that has a string interpolation segment at its end
```